src/main/java/org/testng/TestRunner.java

```diff
@@ -961,7 +975,16 @@ public ISuite getSuite() {
 
   @Override
   public ITestNGMethod[] getAllTestMethods() {
-    return m_allTestMethods;
+    if (getTest().isJUnit()) {
+      //This is true only when we are running JUnit mode
+      return m_allJunitTestMethods;
+    }
+    if (Objects.nonNull(testMethodsContainer) && testMethodsContainer.hasItems()) {
```

Comment:
If testMethodsContainer is null or has no items, why not creating/updating here with computed methods?

Reply:
@juherr - No that is not the intent. Cache should not be re-created in this case.

When running in memory friendly mode and within the context of test execution, the cache will always be valid and it gets explicitly nullified after the test execution is complete. So all invokers of `ITestContext.getAllTestMethods()` will resort to explicit computation (the only situation when this would be invoked is during reporting phase and I think its ok to calculate all the methods.
in the usual mode (ie., no memory friendly mode) the cache is never null (or) empty. So this is not needed

Reply:
The code is not as clear as your explanations ;)
My point is the cache management should be the easiest possible and be understandable without the knowledge of the assumption you mentioned. That's why I'd like to apply here the default pattern of the cache management.

At least, I think you should add comments because they will help later.